File: src/main/java/me/creepinson/creepinoutils/util/BlockUtils.java
package me.creepinson.creepinoutils.util;

import dev.throwouterror.util.math.Tensor;
import net.minecraft.block.AirBlock;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.Direction;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IWorld;
import net.minecraft.world.World;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.util.LazyOptional;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class BlockUtils {
    public static void breakBlockWithDrop(World world, BlockPos pos) {
        BlockState BlockState = world.getBlockState(pos);
        if (BlockState.getMaterial().isLiquid()) {
            world.setBlockState(pos, Blocks.AIR.getDefaultState(), 3);
        } else {
            world.destroyBlock(pos, true);
        }
    }

    /**
     * Safely checks if the tile entity exists.
     *
     * @return a LazyOptional that you can use to check whether the tile is present.
     */
    public static LazyOptional<? extends TileEntity> getTile(IWorld world, BlockPos pos) {
        TileEntity tile = world.getTileEntity(pos);
        return LazyOptional.of(tile != null && !tile.isRemoved() ? () -> tile : null);
    }

    /**
     * Safely checks if the tile entity exists.
     *
     * @return a LazyOptional that you can use to check whether the tile is present.
     */
    public static LazyOptional<? extends TileEntity> getTile(IWorld world, Tensor pos) {
        return getTile(world, TensorUtils.toBlockPos(pos));
    }

    public static boolean isValid(World world, BlockPos pos) {
        return world.isAreaLoaded(pos, 1);
    }

    public static boolean canReplace(World world, BlockPos pos) {
        return isValid(world, pos) && (world.isAirBlock(pos) || world.getBlockState(pos).getMaterial().isReplaceable());
    }

    public static boolean isDirect(Tensor p, Tensor t) {
        return ((p.x() == t.x() + 1 || p.x() == t.x() - 1) && !(p.y() == t.y() + 1 || p.y() == t.y() - 1)
                && !(p.z() == t.z() + 1 || p.z() == t.z() - 1))
                || (!(p.x() == t.x() + 1 || p.x() == t.x() - 1) && (p.y() == t.y() + 1 || p.y() == t.y() - 1)
                && !(p.z() == t.z() + 1 || p.z() == t.z() - 1))
                || (!(p.x() == t.x() + 1 || p.x() == t.x() - 1) && !(p.y() == t.y() + 1 || p.y() == t.y() - 1)
                && (p.z() == t.z() + 1 || p.z() == t.z() - 1));
    }

    public static boolean isIndirect(Tensor p, Tensor t) {
        return !isDirect(p, t);
    }

    /**
     * Returns a set containing the positions of each tile entity that is found to
     * be connected to the starting position.
     */
    public static Set<BlockPos> getTilesWithCapability(World world, BlockPos startingPosition, Capability<?>... search) {
        Set<BlockPos> set = new HashSet<>();
        getTilesWithCapabilityRecursive(set, world, startingPosition, null, search);
        return set;
    }

    public static void getTilesWithCapabilityRecursive(Set<BlockPos> done, World world, BlockPos start, Direction from,
                                                       Capability<?>... search) {
        for (Direction side : Direction.values()) {
            if (side == from)
                continue;
            getTile(world, start).ifPresent(tile -> {
                for (Capability<?> c : search)
                    if (tile.getCapability(c, from).isPresent())
                        done.add(start);

                if (done.contains(start.offset(side)))
                    getTilesWithCapabilityRecursive(done, world, start.offset(side), side.getOpposite(), search);
            });
        }
    }

    /**
     * Returns a set containing the positions of each tile entity that is found to
     * be connected to the starting position.
     */
    public static Set<BlockPos> getTiles(World world, BlockPos startingPosition, Class<?>... search) {
        Set<BlockPos> set = new HashSet<>();
        getTilesRecursive(set, world, startingPosition, null, search);
        return set;
    }

    public static void getTilesRecursive(Set<BlockPos> done, World world, BlockPos start, Direction from,
                                         Class<?>... search) {
        for (Direction side : Direction.values()) {
            if (side == from)
                continue;
            getTile(world, start).ifPresent(tile -> {
                for (Class<?> c : search)
                    if (c.isInstance(tile)) done.add(start);

                if (done.contains(start.offset(side)))
                    getTilesRecursive(done, world, start.offset(side), side.getOpposite(), search);
            });
        }
    }

    /**
     * Returns a set containing the positions of each tile entity that is found to
     * be connected to the starting position.
     */
    public static Set<BlockPos> getBlocks(World world, BlockPos startingPosition, Class<?>... search) {
        Set<BlockPos> set = new HashSet<>();
        getBlocksRecursive(set, world, startingPosition, null, search);
        return set;
    }

    public static void getBlocksRecursive(Set<BlockPos> done, World world, BlockPos start, Direction from,
                                          Class<?>... search) {
        for (Direction side : Direction.values()) {
            if (side == from)
                continue;
            Block b = world.getBlockState(start).getBlock();
            for (Class<?> c : search)
                if (c.isInstance(b))
                    done.add(start);

            if (done.contains(start.offset(side)))
                getBlocksRecursive(done, world, start.offset(side), side.getOpposite(), search);
        }
    }

    /**
     * @param pos    The starting position
     * @param radius The radius to expand from
     * @return A list that contains each block POSITION within the radius
     */
    public static List<BlockPos> getNearbyBlocks(BlockPos pos, int radius) {
        List<BlockPos> scanResult = new ArrayList<BlockPos>();
        for (int x = pos.getX() - radius; x <= pos.getX() + radius; x++)
            for (int y = pos.getY() - radius; y <= pos.getY() + radius; y++)
                for (int z = pos.getZ() - radius; z <= pos.getZ() + radius; z++)
                    scanResult.add(new BlockPos(x, y, z));

        return scanResult;
    }

    /**
     * @param pos    The starting position
     * @param radius The radius to expand from
     * @return A list that contains each block STATE within the radius. Compared to
     * the other getNearbyBlocks method, this one returns a list of
     * blockstates to make it easier to get each block's information.
     */
    public static List<BlockState> getNearbyBlocks(World world, BlockPos pos, int radius) {
        List<BlockState> scanResult = new ArrayList<BlockState>();
        List<BlockPos> scans = getNearbyBlocks(pos, radius);
        for (BlockPos scanPos : scans)
            scanResult.add(world.getBlockState(scanPos));

        return scanResult;
    }

    /**
     * @param pos    The starting position
     * @param radius The radius to expand from
     * @return A list that contains each block STATE within the radius. Compared to
     * the other getNearbyBlocks method, this one returns a list of
     * blockstates to make it easier to get each block's information.
     */
    public static List<BlockState> getNearbyBlocks(World world, Tensor pos, int radius) {
        List<BlockState> scanResult = new ArrayList<BlockState>();
        List<BlockPos> scans = getNearbyBlocks(TensorUtils.toBlockPos(pos), radius);
        for (BlockPos scanPos : scans)
            scanResult.add(world.getBlockState(scanPos));

        return scanResult;
    }

    public static Direction getNeighborDirection(BlockPos pos, BlockPos neighbor) {
        int dx = pos.getX() - neighbor.getX();

        if (dx == 0) {
            int dz = pos.getZ() - neighbor.getZ();
            if (dz == 0) {
                int dy = pos.getY() - neighbor.getY();
                if (dy >= 1)
                    return Direction.DOWN;
                else
                    return Direction.UP;
            } else if (dz >= 1)
                return Direction.NORTH;
            else
                return Direction.SOUTH;
        } else if (dx >= 1)
            return Direction.WEST;
        else
            return Direction.EAST;
    }

    public static BlockState getState(ItemStack stack) {
        Block block = Block.getBlockFromItem(stack.getItem());
        if (!(block instanceof AirBlock))
            return block.getDefaultState();
        return Blocks.AIR.getDefaultState();
    }
}

File: src/main/kotlin/com/theoparis/creepinoutils/util/BlockUtils.kt
package com.theoparis.creepinoutils.util

import net.minecraft.block.AirBlock
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.Blocks
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.Direction
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IWorld
import net.minecraft.world.World
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.util.LazyOptional
import net.minecraftforge.common.util.NonNullSupplier
import java.util.*

object BlockUtils {
    fun breakBlockWithDrop(world: World, pos: BlockPos) {
        val blockState = world.getBlockState(pos)
        if (blockState.material.isLiquid) {
            world.setBlockState(pos, Blocks.AIR.defaultState, 3)
        } else {
            world.destroyBlock(pos, true)
        }
    }

    /**
     * Safely checks if the tile entity exists.
     *
     * @return a LazyOptional that you can use to check whether the tile is present.
     */
    fun getTile(world: World, pos: BlockPos): LazyOptional<out TileEntity> {
        val tile = world.getTileEntity(pos)
        return LazyOptional.of(if (tile != null && !tile.isRemoved) NonNullSupplier { tile } else null)
    }

    fun isValid(world: World, pos: BlockPos): Boolean {
        return world.isAreaLoaded(pos, 1)
    }

    fun canReplace(world: World, pos: BlockPos): Boolean {
        return isValid(world, pos) && (world.isAirBlock(pos) || world.getBlockState(pos).material.isReplaceable)
    }

    fun isDirect(p: BlockPos, t: BlockPos): Boolean {
        return (((p.x == t.x + 1 || p.x == t.x - 1) && !(p.y == t.y + 1 || p.y == t.y - 1)
                && !(p.z == t.z + 1 || p.z == t.z - 1))
                || (!(p.x == t.x + 1 || p.x == t.x - 1) && (p.y == t.y + 1 || p.y == t.y - 1)
                && !(p.z == t.z + 1 || p.z == t.z - 1))
                || (!(p.x == t.x + 1 || p.x == t.x - 1) && !(p.y == t.y + 1 || p.y == t.y - 1)
                && (p.z == t.z + 1 || p.z == t.z - 1)))
    }

    /**
     * Returns a set containing the positions of each tile entity that is found to
     * be connected to the starting position.
     */
    fun getTilesWithCapability(world: World, startingPosition: BlockPos, vararg search: Capability<*>): Set<BlockPos> {
        val set: MutableSet<BlockPos> = HashSet()
        getTilesWithCapabilityRecursive(set, world, startingPosition, null, *search)
        return set
    }

    fun getTilesWithCapabilityRecursive(
        done: MutableSet<BlockPos>, world: World, start: BlockPos, from: Direction?,
        vararg search: Capability<*>
    ) {
        for (side in Direction.values()) {
            if (side == from) continue
            getTile(world, start).ifPresent { tile: TileEntity ->
                for (c in search) if (tile.getCapability(c, from).isPresent) done.add(start)
                if (done.contains(start.offset(side))) getTilesWithCapabilityRecursive(
                    done,
                    world,
                    start.offset(side),
                    side.opposite,
                    *search
                )
            }
        }
    }

    /**
     * Returns a set containing the positions of each tile entity that is found to
     * be connected to the starting position.
     */
    fun getTiles(world: World, startingPosition: BlockPos, vararg search: Class<*>): Set<BlockPos> {
        val set: MutableSet<BlockPos> = HashSet()
        getTilesRecursive(set, world, startingPosition, null, *search)
        return set
    }

    fun getTilesRecursive(
        done: MutableSet<BlockPos>, world: World, start: BlockPos, from: Direction?,
        vararg search: Class<*>
    ) {
        for (side in Direction.values()) {
            if (side == from) continue
            getTile(world, start).ifPresent { tile: TileEntity? ->
                for (c in search) if (c.isInstance(tile)) done.add(start)
                if (done.contains(start.offset(side))) getTilesRecursive(
                    done,
                    world,
                    start.offset(side),
                    side.opposite,
                    *search
                )
            }
        }
    }

    /**
     * Returns a set containing the positions of each tile entity that is found to
     * be connected to the starting position.
     */
    fun getBlocks(world: World, startingPosition: BlockPos, vararg search: Class<*>): Set<BlockPos> {
        val set: MutableSet<BlockPos> = HashSet()
        getBlocksRecursive(set, world, startingPosition, null, *search)
        return set
    }

    fun getBlocksRecursive(
        done: MutableSet<BlockPos>, world: World, start: BlockPos, from: Direction?,
        vararg search: Class<*>
    ) {
        for (side in Direction.values()) {
            if (side == from) continue
            val b = world.getBlockState(start).block
            for (c in search) if (c.isInstance(b)) done.add(start)
            if (done.contains(start.offset(side))) getBlocksRecursive(
                done,
                world,
                start.offset(side),
                side.opposite,
                *search
            )
        }
    }

    /**
     * @param pos    The starting position
     * @param radius The radius to expand from
     * @return A list that contains each block POSITION within the radius
     */
    fun getNearbyBlocks(pos: BlockPos, radius: Int): List<BlockPos> {
        val scanResult: MutableList<BlockPos> = ArrayList()
        for (x in pos.x - radius..pos.x + radius) for (y in pos.y - radius..pos.y + radius) for (z in pos.z - radius..pos.z + radius) scanResult.add(
            BlockPos(x, y, z)
        )
        return scanResult
    }

    /**
     * @param pos    The starting position
     * @param radius The radius to expand from
     * @return A list that contains each block STATE within the radius. Compared to
     * the other getNearbyBlocks method, this one returns a list of
     * block states to make it easier to get each block's information.
     */
    fun getNearbyBlocks(world: World, pos: BlockPos, radius: Int): List<BlockState> {
        val scanResult: MutableList<BlockState> = ArrayList()
        val scans = getNearbyBlocks(pos, radius)
        for (scanPos in scans) scanResult.add(world.getBlockState(scanPos))
        return scanResult
    }

    fun getNeighborDirection(pos: BlockPos, neighbor: BlockPos): Direction {
        val dx = pos.x - neighbor.x
        return if (dx == 0) {
            val dz = pos.z - neighbor.z
            if (dz == 0) {
                val dy = pos.y - neighbor.y
                if (dy >= 1) Direction.DOWN else Direction.UP
            } else if (dz >= 1) Direction.NORTH else Direction.SOUTH
        } else if (dx >= 1) Direction.WEST else Direction.EAST
    }

    fun getState(stack: ItemStack): BlockState {
        val block = Block.getBlockFromItem(stack.item)
        return if (block !is AirBlock) block.defaultState else Blocks.AIR.defaultState
    }
}